cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.batch;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.format.StructuredRecordStringConverter;
+import org.apache.hadoop.io.WritableComparable;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * Wrapper around a {@link StructuredRecord} so that it can be used as the output key and/or value of a mapper.
+ * This is not very performant and must be improved soon (CDAP-5347).
+ */
+public class StructuredRecordWritable implements WritableComparable<StructuredRecordWritable> {
+  private StructuredRecord record;
+
+  // required by Hadoop
+  @SuppressWarnings("unused")
+  public StructuredRecordWritable() {
+  }
+
+  public StructuredRecordWritable(StructuredRecord record) {
+    this.record = record;
+  }
+
+  public void set(StructuredRecord record) {
+    this.record = record;
+  }
+
+  public StructuredRecord get() {
+    return record;
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  @Override
+  public void write(DataOutput out) throws IOException {
+    String schemaStr = record.getSchema().toString();
+    String recordStr = StructuredRecordStringConverter.toJsonString(record);
+    out.writeInt(schemaStr.length());
+    out.write(Bytes.toBytes(schemaStr));
+    out.writeInt(recordStr.length());
+    out.write(Bytes.toBytes(recordStr));
+  }
```

Comment:
Why are we serializing to JSON? That is very verbose and will result in many bytes to shuffle. The JSON essentially repeats the schema... Ideally we would serialize the schema only once, to the config (it is the same for all records, right?) and then use an AVRO-like encoding for each record.